### ISSUE TITLE
MS Sensor Multiplexer

### ISSUE THREAD

**Issue by donei34** [ev3go/ev3dev#81]
Hi,

I was wondering how the Mindsensors Sensor Multiplexer could be used in this language, including:

- How do I create a lego port with an incorrect / undefined driver (the device must be set for each port due to lack of autodetection)?
- What sort of address for the port do I need to use?
- How do I get individual values from the multiplexed ports? Is it like usual?

Thanks in advance

**Comment by kortschak** [comment]
First up, what device (note that that device does not work on BrickPI AFAIK) are you using this on and what have you done so far?

The ev3go API is a thin layer over the ev3dev sysfs control for EV3 etc devices, so if you have got the smux to work on the command line (this is how I explore when setting things up), then it's fairly straight forward to convert that to ev3go. The details for the MS smux are [here for jessie](http://docs.ev3dev.org/projects/lego-linux-drivers/en/ev3dev-jessie/muxs.html#ms-ev3-smux-port-mode-info) and [here for stretch](http://docs.ev3dev.org/projects/lego-linux-drivers/en/ev3dev-stretch/muxs.html#ms-ev3-smux-port-mode-info). I don't have the device, so I cannot give you a worked solution, however, you can see how LEGO ports work [here for stretch](http://docs.ev3dev.org/projects/lego-linux-drivers/en/ev3dev-stretch/ports.html). I would recommend that you try setting up the connections from a shell first to see what you need to do to get the system you want set up. Then I can help translating that to the ev3go API.

**Comment by dlech** [comment]
> What sort of address for the port do I need to use?

You can find the address in the Device Browser on the EV3.


>     * How do I create a lego port with an incorrect / undefined driver (the device must be set for each port due to lack of autodetection)?

https://github.com/ev3go/ev3dev/blob/master/lego_port.go#L160

A short delay is probably needed after calling this before the new sensor device appears.


>     * How do I get individual values from the multiplexed ports? Is it like usual?

The only unusual things are the delay and the address, both already mentioned above.

**Comment by donei34** [comment]
Hi, 
Thank you both for your answers. So, I haven't currently got anything in go, but using an ev3, two sensor multiplexers, six sensors and python I have got them working. 
As such, I don't actually have an idea as to what happens in the shell / fs. I guess I'll just try out what I can and report back!
Thanks

**Comment by kortschak** [comment]
From a brief look at the driver source I think you want something like this (it appears that the smux is automatically detected) [untested]:
```
mux, err := ev3dev.LegoPortFor("ev3-ports:in1", "ms-ev3-smux-port")
handle(err)
addr, err := ev3dev.AddressOf(mux)
handle(err)

p1, err := ev3dev.LegoPortFor(addr+":mux1", "")
handle(err)
err = p1.SetDevice("lego-ev3-us").Err()
handle(err)
s1, err := ev3dev.SensorFor(addr+":mux1", "")
handle(err)

<use s1>
```
Does this look reasonable @dlech?

**Comment by donei34** [comment]
Using the above code I get the following errors:
```
<nil>
<nil>
ev3dev: could not find device for driver "" on port ev3-ports:in1:mux1
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0xa7a40]

goroutine 1 [running]:
github.com/ev3go/ev3dev.(*LegoPort).SetDevice(0x0, 0xd56d7, 0xb, 0x47)
        C:/Users/Dan/go/src/github.com/ev3go/ev3dev/lego_port.go:161 +0x14
main.main()
        C:/Users/Dan/Documents/vscode-hello-go-master/move.go:17 +0x1a0
```

**Comment by donei34** [comment]
Hi, 

Thanks for all of your hope, but I managed to get it working by using my above code (basically direct translation from python):

package main

```
import (
	"fmt"
	"log"

	"github.com/ev3go/ev3dev"
)

func main() {
	port, err := ev3dev.LegoPortFor("i2c-legoev33:i2c80:mux1", "")
	port.SetDevice("lego-ev3-us")
	s1, err := ev3dev.SensorFor("i2c-legoev33:i2c80:mux1", "")
	fmt.Println(err)
	fmt.Println(s1.Value(0))
	log.Println(s1.Value(0))
}

```
Thus I would consider this solved!!!

**Comment by donei34** [comment]
Not sure sorry. TBH, this is my first go program ever, so I have no idea how to handle errors. Is there anything else you'd like me to try / test for you?

**Comment by dlech** [comment]
> Does writing "ms-ev3-smux-port" to `port<N>/driver` (when that port has a MS smux attached) cause a new set of paths to appear?

Not to `driver`, but rather to `set_device`.

> Clearly this doesn't work, but why?

There seems to be a bug in in the kernel drivers that the `address` attributes for I2C devices uses the "wrong" prefix. It looks like it is using `i2c-legoev33:i2c80` when it should of been `ev3-ports:in1:i2c80`. So, I think if you swap that out in the code you gave, then it would work.

**Comment by kortschak** [comment]
@donei34 For error handling, you should check whether functions that return errors have returned a non-nil error and then react accordingly. Nearly always when an non-nil error is returned the other values will either not be usable or will not be usable correctly. In the case of the ev3dev package, they are returned nil and so you see the nil pointer dereference panic.

@dlech Thanks. That `address` does look weird. Do you have any idea where it's coming from?

**Comment by kortschak** [comment]
That's unfortunate. Is there anything that can be done there?

**Comment by dlech** [comment]
Yes, I'm sure it could be fixed.

**Comment by dlech** [comment]
FYI, I've fixed this in kernel ev3dev-2.3.0, so if you upgrade your kernel they you will need to change the address names.

**Comment by kortschak** [comment]
Thanks.
